test/parallel/test-vm-object-keys.js

```diff
@@ -0,0 +1,43 @@
+'use strict';
+require('../common');
+const assert = require('assert');
+const vm = require('vm');
+
+// Regression test for https://github.com/nodejs/node/issues/22723.
+
+const kFoo = Symbol('kFoo');
+
+const test = {
+  'not': 'empty',
+  '0': 0,
+  '0.5': 0.5,
+  '1': 1,
+  '-1': -1,
+  [kFoo]: kFoo
+};
+vm.createContext(test);
+const proxied = vm.runInContext('this', test);
+
+assert.deepStrictEqual(
+  Object.keys(proxied).sort(),
+  ['0', '1', 'not', '0.5', '-1'].sort());
```

Comment:
Is the sorting necessary? `Object.keys()` and `Object.getOwnPropertyNames()` should always return the properties in the order you already expect them to be. Everything else is likely a bug.

Reply:
It fails if I don’t do this, even though this is the order in which they were declared. I’m not sure on the “why” of that, though. (i.e. this is also a question for @nodejs/vm)

Reply:
By spec the return value is:

1. all indices (integers >= 0) in ascending order
2. all other non-symbol properties in their insertion order
3. all symbol properties in their insertion order

So if something else is returned, we should probably look into that.

Reply:
Would you be so kind and add a comment that this is against the spec and should be fixed?

Reply:
@bridgear does b9ec6b825844 sound okay?

Reply:
Yes :)